pkg/kubelet/cm/devicemanager/manager.go

```diff
@@ -588,20 +588,45 @@ func (m *ManagerImpl) writeCheckpoint() error {
 	return nil
 }
 
-// Reads device to container allocation information from disk, and populates
-// m.allocatedDevices accordingly.
-func (m *ManagerImpl) readCheckpoint() error {
+func (m *ManagerImpl) restoreCheckpoint() (checkpoint.DeviceManagerCheckpoint, error) {
 	registeredDevs := make(map[string][]string)
 	devEntries := make([]checkpoint.PodDevicesEntry, 0)
 	cp := checkpoint.New(devEntries, registeredDevs)
 	err := m.checkpointManager.GetCheckpoint(kubeletDeviceManagerCheckpoint, cp)
+	return cp, err
+}
+
+func (m *ManagerImpl) restoreCheckpointV1() (checkpoint.DeviceManagerCheckpoint, error) {
+	registeredDevs := make(map[string][]string)
+	devEntries := make([]checkpoint.PodDevicesEntryV1, 0)
+	cp := checkpoint.NewV1(devEntries, registeredDevs)
+	err := m.checkpointManager.GetCheckpoint(kubeletDeviceManagerCheckpoint, cp)
+	return cp, err
+}
+
+// Reads device to container allocation information from disk, and populates
+// m.allocatedDevices accordingly.
+func (m *ManagerImpl) readCheckpoint() error {
+	// the vast majority of time we restore a compatible checkpoint, so we try
+	// the current version first. Trying to restore older format checkpoints is
+	// relevant only in the kubelet upgrade flow, which happens once in a
+	// (long) while.
+	cp, err := m.restoreCheckpoint()
 	if err != nil {
 		if err == errors.ErrCheckpointNotFound {
+			// no point in trying anything else
 			klog.InfoS("Failed to retrieve checkpoint", "checkpoint", kubeletDeviceManagerCheckpoint, "err", err)
 			return nil
 		}
-		return err
+
+		// one last try: maybe it's a old format checkpoint?
+		cp, err = m.restoreCheckpointV1()
+		if err != nil {
+			return err
```

Comment:
I'm not sure we want to return the v1 error - Or at least not only the v1 error unless we also log the current version error

Reply:
logging the two errors looks like a nice idea to me

Reply:
actually decided to return the previous error (not the v1 error) because I couldn't find yet a good way to wrap two of them. I'm not very very fond of my current approach - even though I'd still call an improvement, so totally open for further discussion.

Reply:
That works I think, wrapping multi errors is always kinda uncomfortable